File: repurpose-agent/video/scripts/transcribe.py
#!/usr/bin/env python3
"""
transcribe.py — transcribe a video/audio file with Deepgram and write
timestamped, speaker-labeled markdown next to the source file.

Usage:
    DEEPGRAM_API_KEY=... python3 transcribe.py /path/to/videos/webinar.mp4

Output:
    /path/to/videos/webinar.md   (markdown transcript)

Requires:
    - ffmpeg
    - python3 (stdlib only; no third-party packages)
    - DEEPGRAM_API_KEY environment variable (a Deepgram API key with STT scope)

This calls the Deepgram pre-recorded HTTP API directly (nova-3, diarized,
smart-format). It does not depend on the `dg` CLI.

Speaker labels are emitted generically as **Mike Shima/1/2**; map them to real
names in a post-step (Deepgram assigns speaker indices per run).
"""

import json
import os
import subprocess
import sys
import tempfile
import urllib.error
import urllib.parse
import urllib.request

DEEPGRAM_URL = "https://api.deepgram.com/v1/listen"


def fmt_ts(seconds: float) -> str:
    """Convert seconds (float) to [HH:MM:SS] or [MM:SS]."""
    seconds = int(seconds)
    h, rem = divmod(seconds, 3600)
    m, s = divmod(rem, 60)
    if h:
        return f"[{h:02d}:{m:02d}:{s:02d}]"
    return f"[{m:02d}:{s:02d}]"


def main():
    if len(sys.argv) < 2:
        print("Usage: python3 transcribe.py <video-or-audio-file>", file=sys.stderr)
        sys.exit(1)

    src = os.path.abspath(os.path.expanduser(sys.argv[1]))
    if not os.path.isfile(src):
        print(f"File not found: {src}", file=sys.stderr)
        sys.exit(1)

    api_key = os.environ.get("DEEPGRAM_API_KEY")
    if not api_key:
        print(
            "DEEPGRAM_API_KEY is not set. Export a Deepgram API key with STT "
            "scope, e.g.:\n    export DEEPGRAM_API_KEY=your_key_here",
            file=sys.stderr,
        )
        sys.exit(1)

    stem = os.path.splitext(os.path.basename(src))[0]
    if len(sys.argv) >= 3:
        out_dir = os.path.abspath(os.path.expanduser(sys.argv[2]))
        os.makedirs(out_dir, exist_ok=True)
        out_md = os.path.join(out_dir, stem + ".md")
    else:
        out_md = os.path.join(os.path.dirname(src), stem + ".md")

    # 1. Extract audio to a compact temp mp3 (mono, 16kHz — small upload,
    #    no meaningful quality loss for speech).
    tmp_audio = tempfile.NamedTemporaryFile(suffix=".mp3", delete=False).name
    print(f"Extracting audio -> {tmp_audio}", file=sys.stderr)
    subprocess.run(
        ["ffmpeg", "-i", src, "-vn", "-ac", "1", "-ar", "16000", "-b:a", "48k",
         tmp_audio, "-y"],
        check=True,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )

    # 2. POST to the Deepgram pre-recorded HTTP API (nova-3, diarized,
    #    smart-format). Stdlib urllib only — no `dg` CLI, no `requests`.
    print("Transcribing with Deepgram (nova-3, diarized)...", file=sys.stderr)
    query = urllib.parse.urlencode({
        "model": "nova-3",
        "diarize": "true",
        "smart_format": "true",
        "punctuate": "true",
    })
    with open(tmp_audio, "rb") as fh:
        audio_bytes = fh.read()
    req = urllib.request.Request(
        f"{DEEPGRAM_URL}?{query}",
        data=audio_bytes,
        method="POST",
        headers={
            "Authorization": f"Token {api_key}",
            "Content-Type": "audio/mpeg",
        },
    )
    try:
        with urllib.request.urlopen(req, timeout=600) as resp:
            data = json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", "replace")[:500]
        print(f"Deepgram API error {e.code}: {body}", file=sys.stderr)
        os.unlink(tmp_audio)
        sys.exit(1)
    except urllib.error.URLError as e:
        print(f"Could not reach Deepgram: {e.reason}", file=sys.stderr)
        os.unlink(tmp_audio)
        sys.exit(1)

    # 3. Walk words, grouping consecutive words by speaker into turns.
    #    Within each turn, break the text into sentences and keep the start
    #    timestamp of every sentence — this is what lets downstream tooling
    #    pin a stat/quote to the exact second instead of estimating its
    #    position inside a multi-minute speaker block.
    words = data["results"]["channels"][0]["alternatives"][0]["words"]
    if not words:
        print("No words returned — check the audio file.", file=sys.stderr)
        sys.exit(1)

    SENTENCE_END = (".", "?", "!")

    # Each turn is (speaker, turn_start, [ (sentence_start, sentence_text), ... ])
    turns = []
    cur_spk = None
    cur_turn_start = None
    cur_sentences = []          # closed sentences in the current turn
    sent_words = []             # words accumulating into the current sentence
    sent_start = None           # start timestamp of the current sentence

    def close_sentence():
        if sent_words:
            cur_sentences.append((sent_start, " ".join(sent_words)))

    def close_turn():
        close_sentence()
        if cur_sentences:
            turns.append((cur_spk, cur_turn_start, list(cur_sentences)))

    for w in words:
        spk = w.get("speaker", 0)
        # smart_format puts the cleaned token in punctuated_word when available
        token = w.get("punctuated_word", w["word"])

        if spk != cur_spk:
            close_turn()
            cur_spk = spk
            cur_turn_start = w["start"]
            cur_sentences = []
            sent_words = []
            sent_start = None

        if not sent_words:
            sent_start = w["start"]
        sent_words.append(token)

        # Close the sentence on terminal punctuation.
        if token.rstrip('"\')]').endswith(SENTENCE_END):
            close_sentence()
            sent_words = []
            sent_start = None

    close_turn()

    # 4. Write markdown — speaker header carries the turn start; every
    #    sentence line is prefixed with its own [MM:SS] timestamp.
    title = stem
    lines = [f"# Transcript: {title}", ""]
    for spk, turn_start, sentences in turns:
        lines.append(f"**Speaker {spk}** {fmt_ts(turn_start)}")
        lines.append("")
        for sent_start, text in sentences:
            lines.append(f"{fmt_ts(sent_start)} {text}")
        lines.append("")

    with open(out_md, "w") as f:
        f.write("\n".join(lines))

    os.unlink(tmp_audio)
    print(f"\nDone -> {out_md}", file=sys.stderr)
    print(out_md)


if __name__ == "__main__":
    main()

File: src/lib/transcribe.ts
// Deepgram transcription for the repurpose pipeline.
//
// Uses Deepgram's URL-based prerecorded API: we hand Deepgram a presigned GET
// url to the uploaded video and Deepgram extracts the audio server-side, so the
// (multi-GB) media never passes through this process. Then we port the
// word-grouping logic from repurpose-agent/video/scripts/transcribe.py: group
// consecutive words by speaker into turns, break each turn into sentences, and
// stamp EVERY sentence with its own [MM:SS] timestamp. That per-sentence stamp
// is what lets the social generator pin a clip to the exact second a stat or
// quote is spoken instead of estimating inside a multi-minute speaker block.
//
// Output markdown is byte-compatible with transcribe.py so the create-social
// skill's timestamp-reading instructions apply unchanged.

const DEEPGRAM_URL = "https://api.deepgram.com/v1/listen";

interface DeepgramWord {
  word: string;
  punctuated_word?: string;
  start: number;
  end: number;
  speaker?: number;
}

interface DeepgramResponse {
  results?: {
    channels?: Array<{
      alternatives?: Array<{ words?: DeepgramWord[] }>;
    }>;
  };
}

/** Convert seconds to [HH:MM:SS] or [MM:SS] — matches transcribe.py fmt_ts. */
function fmtTs(seconds: number): string {
  const total = Math.floor(seconds);
  const h = Math.floor(total / 3600);
  const m = Math.floor((total % 3600) / 60);
  const s = total % 60;
  const pad = (n: number) => String(n).padStart(2, "0");
  if (h) return `[${pad(h)}:${pad(m)}:${pad(s)}]`;
  return `[${pad(m)}:${pad(s)}]`;
}

const SENTENCE_END = [".", "?", "!"];

function endsSentence(token: string): boolean {
  const trimmed = token.replace(/["')\]]+$/, "");
  return SENTENCE_END.some((p) => trimmed.endsWith(p));
}

interface Turn {
  speaker: number;
  turnStart: number;
  sentences: Array<{ start: number; text: string }>;
}

// Port of transcribe.py lines 118–184: walk words, group by speaker into turns,
// break turns into sentences carrying their own start timestamp.
function wordsToMarkdown(title: string, words: DeepgramWord[]): string {
  const turns: Turn[] = [];

  let curSpk: number | null = null;
  let curTurnStart = 0;
  let curSentences: Turn["sentences"] = [];
  let sentWords: string[] = [];
  let sentStart: number | null = null;

  const closeSentence = () => {
    if (sentWords.length && sentStart != null) {
      curSentences.push({ start: sentStart, text: sentWords.join(" ") });
    }
  };
  const closeTurn = () => {
    closeSentence();
    if (curSentences.length && curSpk != null) {
      turns.push({ speaker: curSpk, turnStart: curTurnStart, sentences: curSentences });
    }
  };

  for (const w of words) {
    const spk = w.speaker ?? 0;
    const token = w.punctuated_word ?? w.word;

    if (spk !== curSpk) {
      closeTurn();
      curSpk = spk;
      curTurnStart = w.start;
      curSentences = [];
      sentWords = [];
      sentStart = null;
    }

    if (!sentWords.length) sentStart = w.start;
    sentWords.push(token);

    if (endsSentence(token)) {
      closeSentence();
      sentWords = [];
      sentStart = null;
    }
  }
  closeTurn();

  const lines: string[] = [`# Transcript: ${title}`, ""];
  for (const turn of turns) {
    lines.push(`**Speaker ${turn.speaker}** ${fmtTs(turn.turnStart)}`);
    lines.push("");
    for (const sent of turn.sentences) {
      lines.push(`${fmtTs(sent.start)} ${sent.text}`);
    }
    lines.push("");
  }
  return lines.join("\n");
}

const RETRYABLE = new Set([429, 500, 502, 503, 504]);

/**
 * Transcribe the media at `url` (a presigned GET url Deepgram fetches itself)
 * into timestamped, speaker-diarized markdown. `title` becomes the H1 and is the
 * only cosmetic difference from the Python script's output.
 */
export async function transcribeFromUrl(url: string, title: string): Promise<string> {
  const apiKey = process.env.DEEPGRAM_API_KEY;
  if (!apiKey) {
    throw new Error("DEEPGRAM_API_KEY is not set. Add a Deepgram STT key to .env.");
  }

  const params = new URLSearchParams({
    model: "nova-3",
    diarize: "true",
    smart_format: "true",
    punctuate: "true",
  });

  let lastErr: unknown;
  for (let attempt = 0; attempt < 3; attempt++) {
    try {
      const resp = await fetch(`${DEEPGRAM_URL}?${params}`, {
        method: "POST",
        headers: {
          Authorization: `Token ${apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ url }),
      });

      if (!resp.ok) {
        const text = (await resp.text()).slice(0, 500);
        if (RETRYABLE.has(resp.status) && attempt < 2) {
          lastErr = new Error(`Deepgram ${resp.status}: ${text}`);
          await new Promise((r) => setTimeout(r, 1000 * 2 ** attempt));
          continue;
        }
        throw new Error(`Deepgram API error ${resp.status}: ${text}`);
      }

      const data = (await resp.json()) as DeepgramResponse;
      const words = data.results?.channels?.[0]?.alternatives?.[0]?.words;
      if (!words || words.length === 0) {
        throw new Error("Deepgram returned no words — check the source media.");
      }
      return wordsToMarkdown(title, words);
    } catch (err) {
      lastErr = err;
      // Network-level failure: retry a couple of times, then give up.
      if (attempt < 2) {
        await new Promise((r) => setTimeout(r, 1000 * 2 ** attempt));
        continue;
      }
      throw err;
    }
  }
  throw lastErr instanceof Error ? lastErr : new Error(String(lastErr));
}

/**
 * Apply a Speaker N → real-name rename to a transcript. The diarized transcript
 * labels turns "**Speaker 0**", "**Speaker 1**", … The UI collects a map from
 * those labels to real names; this rewrites every occurrence so quotes attribute
 * correctly downstream. Whole-label match only, so a name can't partially hit.
 */
export function applySpeakerMap(
  transcript: string,
  speakerMap: Record<string, string>,
): string {
  let out = transcript;
  for (const [label, name] of Object.entries(speakerMap)) {
    const clean = name.trim();
    if (!clean) continue;
    // Match the bolded header form "**Speaker 0**" exactly.
    const re = new RegExp(`\\*\\*${escapeRegExp(label)}\\*\\*`, "g");
    out = out.replace(re, `**${clean}**`);
  }
  return out;
}

function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

/**
 * The distinct speaker labels present in a transcript ("Speaker 0", …), in
 * first-appearance order. Drives the rename form (one input per detected speaker).
 */
export function detectSpeakers(transcript: string): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  const re = /\*\*(Speaker \d+)\*\*/g;
  let m: RegExpExecArray | null;
  while ((m = re.exec(transcript)) !== null) {
    if (!seen.has(m[1])) {
      seen.add(m[1]);
      out.push(m[1]);
    }
  }
  return out;
}
